admin-tools/lambda/src/main/scala/com/gu/mediaservice/BatchIndexHandler.scala

```diff
@@ -0,0 +1,155 @@
+  import InputIdsProvider._
+
+  def processImages()(implicit ec: ExecutionContext): List[Int] = {
+    val stateProgress = scala.collection.mutable.ArrayBuffer[Int]()
+    stateProgress += 0
+    val mediaIds = getMediaIdsBatch
+    Try {
+      println(s"number of mediaIDs to index ${mediaIds.length}, $mediaIds")
+      stateProgress += updateStateToItemsInProgress(mediaIds)
+      val maybeBlobsFuture: Future[List[ImageIdMaybeBlobEntry]] = getMaybeImagesProjectionBlobs(mediaIds)
+      val (foundImageBlobsEntries, notFoundImagesIds) = BatchIndexHandler.partitionToSuccessAndNotFound(maybeBlobsFuture)
+      val imageBlobs = foundImageBlobsEntries.map(_.blob)
+      updateStateToNotFoundImages(notFoundImagesIds).map(stateProgress += _)
+      println(s"prepared json blobs list of size: ${imageBlobs.size}")
+      if (imageBlobs.isEmpty) {
+        println("all was empty terminating current batch")
+        return stateProgress.toList
+      }
+      println("attempting to store blob to s3")
+      val path = putToS3(imageBlobs)
+      val executeBulkIndexMsg = Json.obj(
+        "subject" -> "batch-index",
+        "s3Path" -> path
+      )
+      putToKinensis(executeBulkIndexMsg)
+      stateProgress += updateStateToFinished(foundImageBlobsEntries.map(_.id))
+    } match {
+      case Success(value) => println(s"all good $value")
+      case Failure(exp) =>
+        exp.printStackTrace()
+        println(s"there was a failure, exception: ${exp.getMessage}")
+        stateProgress += resetItemsState(mediaIds)
+        // propagating exception
+        throw exp
+    }
+    println(s"processImages function execution state progress: $stateProgress")
+    stateProgress.toList
+  }
+
+}
+
+class InputIdsProvider(table: Table, batchSize: Int) {
+
+  private val PKField: String = "fileId"
+  private val StateField: String = "fileState"
+
+  def getMediaIdsBatch: List[String] = {
+    println("attempt to get mediaIds batch from dynamo")
+    val scanSpec = new ScanSpec().withFilterExpression(s"$StateField = :sub")
+      .withValueMap(new ValueMap().withNumber(":sub", 0)).withMaxResultSize(batchSize)
+    val mediaIds = table.scan(scanSpec).asScala.toList.map(it => {
+      val json = Json.parse(it.toJSON).as[JsObject]
+      (json \ PKField).as[String]
+    })
+    mediaIds
+  }
+
+  /**
+    * state is used to synchronise multiple overlapping lambda executions, track progress and avoiding repeated operations
+    */
+
+  // used to synchronise situation of other lambda execution will start while previous one is still running
+  def updateStateToItemsInProgress(ids: List[String]): Int = {
+    val state = 1
+    println(s"updating items state to in progress")
+    updateItemsState(ids, state)
+    state
+  }
+
+  // used to track images that were not projected successfully
+  def updateStateToNotFoundImages(notFoundIds: List[String]): Option[Int] = {
+    if (notFoundIds.isEmpty) None else {
+      val state = 2
```

Comment:
Can we create a sum type for `state` to be explicit of their meaning?

Reply:
ok

Reply:
done